A36-xds-for-servers.md

```diff
@@ -193,8 +191,43 @@ required. This is to allow implementations that only support L4 xDS features to
 avoid L7 plumbing and implementation. This has no impact on the resource
 validation and NACKing behavior called for in A39.
 
+If an XdsServer implementation uses RouteConfiguration or supports any HTTP
+filters other than the hard-coded Router, then
+`HttpConnectionManager.route_config` and `HttpConnectionManager.rds` must be
+supported and RouteConfigurations must be validated. RouteConfiguration
+validation logic inherits all previous validations made for client-side usage
```

Comment:
While RDS doesn't/cannot distinguish RouteConfigurations between client-side and server-side, validating/parsing RouteConfiguration requires knowing if it is used for client or server, due to the reason that we need to know if any of the provided HttpFilters is supported for gRPC client or server. As mentioned in [this section of A39](https://github.com/grpc/proposal/blob/master/A39-xds-http-filters.md#top-level-filter-configs): 

> Every filter config in the list must be of a type (as identified by the `type_url` field) that is present in the filter registry. Note that the set of known filter config types will be different on the gRPC client and server; if the `HttpConnectionManager` proto is inside an HTTP API Listener, it will look only at filters registered for the gRPC client, whereas if it is inside a TCP Listener, it will look only at filters registered for the gRPC server. Any unknown config type will cause the Listener resource to be NACKed, unless the `is_optional` field is true, in which case the individual filter will be ignored.

Although that primarily talks about LDS, a RouteConfiguration can be either inlined in an LDS response or received in an RDS response. So I am assuming the rule applies to both. However, we cannot know whether the received RouteConfiguration from RDS is for an API listener that is used for client side or a TCP listener that is used for server side. Unless we are going to (hopefully no) manually bookkeep some coupling inside XdsClient for LDS and RDS resources. Another option is to relax the requirement of NACKing xDS responses with HttpFilters not supported for the specific client-side or server-side use case.

Reply:
> ... Unless we are going to (hopefully no) manually bookkeep some coupling inside XdsClient for LDS and RDS resources. 

That assumes that an RDS resource is uniquely used by just client or server side listener. May be that assumption is true? And if not true, there is no problem?

Reply:
Before this gRFC, we did not have any consideration for RouteConfiguration on server side, no matter it is inlined in an LDS response or received in an RDS response. It [doesn't even look at](https://github.com/grpc/grpc/blob/5a91a513fd8e88f5678bae8ed7ac1e69eaddc757/src/core/ext/xds/xds_api.cc#L2015) the inlined RouteConfiguration in the LDS response when the HttpConnectionManager is from a TCP listener FilterChain. All the RouteConfiguration handling code is based on the assumption that it is for client-side's usage. So this is a new problem.

Reply:
> validating/parsing RouteConfiguration requires knowing if it is used for client or server
>
>> Note that the set of known filter config types will be different on the gRPC client and server; if the HttpConnectionManager proto is inside an HTTP API Listener, it will look only at filters registered for the gRPC client, whereas if it is inside a TCP Listener, it will look only at filters registered for the gRPC server.

That is talking about the _HttpConnectionManager_ which provides the list of filters to be used, which isn't part of RouteConfiguration. We validate the filter list precisely for the use-case (client vs server), but we validate the RouteConfiguration weakly and can only assert that we are aware of the filter (for client _or_ server). But if you put a client-only filter config in RouteConfiguration, there is no way for it to be used on server-side as the server-side validation for HttpConnectionManager will fail _or_ that specific filter config in RouteConfiguration is ignored because the filter isn't actually selected.

It is specifically called out in the RouteConfiguration (`typed_per_filter_config`) part of the gRFC:
> Similarly, gRPC will not fail validation if a map entry uses a config for a filter that is supported on only the gRPC client but is used on the gRPC server, or vice-versa. This is because the xDS client code cannot know whether a given RDS update will be used by a gRPC client or server.